Add lock feature for unique identities

mafesan opened the pull request:
This feature is implemented at DB level for `UniqueIdentity` objects, including mutations for locking and unlocking unique identities (with the corresponding `api` and `db` methods).

The input value for both `lock` and `unlock` mutations (and also for the corresponding API methods), is the `uuid` of the unique identity to be locked or unlocked. 
    
Note that these actions also affect to the unique identity's related objects such as identities, enrollments or its profile.

This `isLocked` new field has been added as a new filter for the query asking for the unique identities.

Moreover, a new error has been added: `LockedIdentityError`. This error shold be raised by all the DB methods modifying UniqueIdentity objects which `is_locked` field is set to `True`. This error includes the `uuid` of the locked unique identity.

---

Mutation examples:
    
```
mutation lockId($uuid: String) {
            lockIdentity(uuid: $uuid) {
              uuid
              uidentity {
                uuid
                isLocked
              }
            }
          }
```
```
mutation unlockId($uuid: String) {
            unlockIdentity(uuid: $uuid) {
              uuid
              uidentity {
                uuid
                isLocked
              }
            }
          }
```

---

This PR is related to #238.

sduenas commented on sortinghat/core/api.py:
```diff
@@ -357,6 +358,68 @@ def move_identity(ctx, from_id, to_uuid):
+
+    uidentity = set_lock_unique_identity_db(trxl, uidentity, True)
+
+    trxl.close()
```
For readability, I think it would be clearer this 
```suggestion
    trxl = TransactionsLog.open('lock_unique_identity', ctx)
    
    uidentity = find_unique_identity(uuid)
    uidentity = set_lock_unique_identity_db(trxl, uidentity, True)
    
    trxl.close()
```

mafesan replied:
Done!

sduenas commented on sortinghat/core/api.py:
```diff
@@ -357,6 +358,68 @@ def move_identity(ctx, from_id, to_uuid):
+
+    uidentity = set_lock_unique_identity_db(trxl, uidentity, False)
+
+    trxl.close()
```
Same as before.

mafesan replied:
Done.

sduenas commented on sortinghat/core/api.py:
```diff
@@ -357,6 +358,68 @@ def move_identity(ctx, from_id, to_uuid):
 
+@django.db.transaction.atomic
+def lock_unique_identity(ctx, uuid):
+    """Lock an unique identity so it cannot be modified
```
```suggestion
    """Lock a unique identity so it cannot be modified.
```

mafesan replied:
Done.

sduenas commented on sortinghat/core/api.py:
```diff
@@ -357,6 +358,68 @@ def move_identity(ctx, from_id, to_uuid):
+
+@django.db.transaction.atomic
+def unlock_unique_identity(ctx, uuid):
+    """Unlock an unique identity so it can be modified
```
```suggestion
    """Unlock a unique identity so it can be modified.
```

mafesan replied:
Done.

sduenas commented on sortinghat/core/db.py:
```diff
@@ -676,6 +698,34 @@ def move_identity(trxl, identity, uidentity):
     return uidentity
 
 
+def set_lock_unique_identity(trxl, uidentity, lock):
```
I think we should define two functions `lock` and `unlock`. These two functions will call this one with the right parameters. Thus, it's easier to understand and functions in higher levels don't need to know what value should given to `lock` in case they want to lock or unlock the identity.

In addition, it would be clearer to rename `lock` value to  `lock_flag` or similar. `lock` sounds you are either locking or passing a lock.

mafesan replied:
Done.

sduenas commented on sortinghat/core/db.py:
```diff
@@ -676,6 +698,34 @@ def move_identity(trxl, identity, uidentity):
 
 
+def set_lock_unique_identity(trxl, uidentity, lock):
+    """Set a lock value for a given unique identity
```
```suggestion
    """Set a lock value for a given unique identity.
```

mafesan replied:
Done.

sduenas commented on sortinghat/core/errors.py:
```diff
@@ -79,3 +80,10 @@ class ClosedTransactionError(BaseError):
+    """Exception raised when performing a change on a locked unique identity"""
+
+    code = CODE_LOCKED_IDENTITY_ERROR
+    message = "%(msg)s"
```
The exception should include the **uuid** of the identity locked.

mafesan replied:
Done.

mafesan commented:
Thank you @sduenas for your comments about the first draft version! This PR is ready for review now.

mafesan commented on sortinghat/core/api.py:
```diff
@@ -357,6 +359,66 @@ def move_identity(ctx, from_id, to_uuid):
+
+    :returns: a unique identity with its locked value updated
+
+    :raises InvalidValueError: raised when `uuid` is `None` or an empty string
```
Done.

mafesan replied:
Done.

mafesan commented on tests/test_schema.py:
```diff
@@ -991,6 +996,115 @@ def test_filter_registry(self):
         self.assertEqual(rol2['start'], '1900-01-01T00:00:00+00:00')
         self.assertEqual(rol2['end'], '2100-01-01T00:00:00+00:00')
 
+    def test_filter_registry_is_locked(self):
```
You are totally right. Now I have fixed this test.

mafesan commented:
Hi @sduenas, I addressed your comments and I have added some additional tests for API methods and mutations performing operations on a locked unique identity. Thank you!
